filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,325 @@
+	// between different policies (i.e. global team memberships)
+	// This not possible due to some limitations in OPA
+	// See https://github.com/open-policy-agent/opa/issues/5707
+}
+
+func NewOpenPolicyAgentFactory() OpenPolicyAgentFactory {
+	return OpenPolicyAgentFactory{}
+}
+
+type OpenPolicyAgentInstanceConfig struct {
+	envoyMetadata          *ext_authz_v3_core.Metadata
+	configTemplate         []byte
+	envoyContextExtensions map[string]string
+}
+
+func WithConfigTemplate(configTemplate []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.configTemplate = configTemplate
+		return nil
+	}
+}
+
+func WithConfigTemplateFile(configTemplateFile string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		var err error
+		cfg.configTemplate, err = os.ReadFile(configTemplateFile)
+		return err
+	}
+}
+
+func WithEnvoyContextExtensions(envoyContextExtensions map[string]string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyContextExtensions = envoyContextExtensions
+		return nil
+	}
+}
+
+func WithEnvoyMetadata(metadata *ext_authz_v3_core.Metadata) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyMetadata = metadata
+		return nil
+	}
+}
+
+func WithEnvoyMetadataBytes(content []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyMetadata = &ext_authz_v3_core.Metadata{}
+		protojson.Unmarshal(content, cfg.envoyMetadata)
+		return nil
+	}
+}
+
+func WithEnvoyMetadataFile(file string) func(*OpenPolicyAgentInstanceConfig) error {
+
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		content, err := os.ReadFile(file)
+		if err != nil {
+			return err
+		}
+		return WithEnvoyMetadataBytes(content)(cfg)
+	}
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetEnvoyMetadata() *ext_authz_v3_core.Metadata {
+	return cfg.envoyMetadata
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetEnvoyContextExtensions() map[string]string {
+	return cfg.envoyContextExtensions
+}
+
+func NewOpenPolicyAgentConfig(opts ...func(*OpenPolicyAgentInstanceConfig) error) (*OpenPolicyAgentInstanceConfig, error) {
+	cfg := OpenPolicyAgentInstanceConfig{}
+
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	if cfg.configTemplate == nil {
+		var err error
+		cfg.configTemplate, err = os.ReadFile(configTemplate())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &cfg, nil
+}
+
+func configTemplate() string {
+	if s, ok := os.LookupEnv("OPA_CONFIG_TEMPLATE_FILE"); ok {
+		return s
+	}
+	return "opaconfig.yaml"
+}
+
+func (factory *OpenPolicyAgentFactory) NewOpenPolicyAgentEnvoyInstance(bundleName string, config OpenPolicyAgentInstanceConfig, filterName string) (*OpenPolicyAgentInstance, error) {
```

Comment:
NewOpenPolicyAgentEnvoyInstance returns OpenPolicyAgentInstanceConfig, so drop also Envoy in this func makes sense